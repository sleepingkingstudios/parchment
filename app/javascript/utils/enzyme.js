import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
