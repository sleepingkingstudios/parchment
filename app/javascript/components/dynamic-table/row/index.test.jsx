import React from 'react';
import { shallow } from 'enzyme';

import DynamicTableRow from './index';

describe('<DynamicTableRow />', () => {
  const columns = [
    { label: true, prop: 'id' },
    { label: true, prop: 'name' },
    { label: true, prop: 'price' },
    { label: true, prop: 'quantity' },
  ];
  const data = {
    id: '00000000-0000-0000-0000-000000000000',
    name: 'Self-sealing Stem Bolt',
    price: '10.00',
    quantity: 50,
  };
  const defaultProps = { columns, data };

  describe('with default props', () => {
    it('should render the table row', () => {
      const rendered = shallow(<DynamicTableRow {...defaultProps} />);

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row');
      expect(rendered).toHaveClassName('row');
    });

    it('should render the row cells', () => {
      const rendered = shallow(<DynamicTableRow {...defaultProps} />);
      const cells = rendered.find('DynamicTableRowCell');

      expect(cells.length).toEqual(columns.length);

      columns.forEach((column, index) => {
        const cell = cells.at(index);
        const { label, prop } = column;

        expect(cell).toHaveProp({ data });
        expect(cell).toHaveProp({ label });
        expect(cell).toHaveProp({ prop });
      });
    });

    it('should match the snapshot', () => {
      const rendered = shallow(<DynamicTableRow {...defaultProps} />);

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with cellProps: object', () => {
    const cellProps = { key: 'value' };

    it('should render the row cells', () => {
      const rendered = shallow(<DynamicTableRow {...defaultProps} cellProps={cellProps} />);
      const cells = rendered.find('DynamicTableRowCell');

      expect(cells.length).toEqual(columns.length);

      columns.forEach((column, index) => {
        const cell = cells.at(index);
        const { label, prop } = column;

        expect(cell).toHaveProp({ cellProps });
        expect(cell).toHaveProp({ data });
        expect(cell).toHaveProp({ label });
        expect(cell).toHaveProp({ prop });
      });
    });
  });

  describe('with resourceName: value', () => {
    const resourceName = 'widgets';

    it('should render the table row', () => {
      const rendered = shallow(<DynamicTableRow {...defaultProps} resourceName={resourceName} />);

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row');
      expect(rendered).toHaveClassName('widgets-table-row');
      expect(rendered).toHaveClassName('row');
    });

    it('should render the row cells', () => {
      const rendered = shallow(<DynamicTableRow {...defaultProps} resourceName={resourceName} />);
      const cells = rendered.find('DynamicTableRowCell');

      expect(cells.length).toEqual(columns.length);

      columns.forEach((column, index) => {
        const cell = cells.at(index);
        const { label, prop } = column;

        expect(cell).toHaveProp({ data });
        expect(cell).toHaveProp({ label });
        expect(cell).toHaveProp({ prop });
        expect(cell).toHaveProp({ resourceName });
      });
    });
  });
});
