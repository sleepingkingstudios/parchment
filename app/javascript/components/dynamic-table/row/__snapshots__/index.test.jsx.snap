// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicTableRow /> with default props should match the snapshot 1`] = `
<div
  className="dynamic-table-row row"
>
  <DynamicTableRowCell
    cellProps={Object {}}
    data={
      Object {
        "id": "00000000-0000-0000-0000-000000000000",
        "name": "Self-sealing Stem Bolt",
        "price": "10.00",
        "quantity": 50,
      }
    }
    header={true}
    key="id"
    label={true}
    prop="id"
    resourceName={null}
    width={null}
  />
  <DynamicTableRowCell
    cellProps={Object {}}
    data={
      Object {
        "id": "00000000-0000-0000-0000-000000000000",
        "name": "Self-sealing Stem Bolt",
        "price": "10.00",
        "quantity": 50,
      }
    }
    header={true}
    key="name"
    label={true}
    prop="name"
    resourceName={null}
    width={null}
  />
  <DynamicTableRowCell
    cellProps={Object {}}
    data={
      Object {
        "id": "00000000-0000-0000-0000-000000000000",
        "name": "Self-sealing Stem Bolt",
        "price": "10.00",
        "quantity": 50,
      }
    }
    header={true}
    key="price"
    label={true}
    prop="price"
    resourceName={null}
    width={null}
  />
  <DynamicTableRowCell
    cellProps={Object {}}
    data={
      Object {
        "id": "00000000-0000-0000-0000-000000000000",
        "name": "Self-sealing Stem Bolt",
        "price": "10.00",
        "quantity": 50,
      }
    }
    header={true}
    key="quantity"
    label={true}
    prop="quantity"
    resourceName={null}
    width={null}
  />
</div>
`;
